first commit with overview page (#14827)
hashicorp/consul: website/content/docs/consul-vs-other/config-management-compare.mdx

---
layout: docs
page_title: Consul compared to other configuration management tools
description: >-
  Chef, Puppet, and other configuration management tools build service discovery mechanisms by querying global state and constructing configuration files on each node during a periodic convergence run.
---

# Consul Compared to Other Configuration Management Tools

> Examples: Chef, Puppet

Reviewer: ```suggestion
**Examples**: Chef, Puppet
```

There are many configuration management tools available. These tools typically focus on static provisioning, while Consul supports dynamic configuration based on service and node state. Both types of configuration are important and work well together. Because Consul offers a number of different capabilities, there are times when its functionality overlaps with other configuration management tools.

For example, Chef and Puppet are configuration management tools that can build service discovery mechanisms. However, they only support configuration information that is static and as a result, the time it takes to implement updates ranges from several minutes to hours depending on the frequency of conversion runs. Additionally, these tools do not include mechanisms to incorporate the system state in the configuration, so they allow load balancers to send traffic to unhealthy nodes, further exacerbating issues. Supporting multiple datacenters is also challenging with these tools, as a central group of servers must manage all datacenters.

> Consul's service discovery layer is specifically designed to dynamically track and respond to the state of the cluster. By using the integrated health checking, Consul can route traffic away from unhealthy nodes, which allows systems and services to gracefully recover. In addition, Consul’s service discovery layer works with Terraform by using Consul-Terraform-Sync (CTS) to automate updates to network infrastructure based on dynamic changes to each service. For example, as services scale up or down, CTS can trigger Terraform and update firewalls or load balancers to reflect the latest changes. Also, because each datacenter runs independently, supporting multiple datacenters is no different than supporting a single datacenter.

Reviewer 2: ```suggestion
Consul's service discovery layer is specifically designed to dynamically track and respond to your service's state. By using the integrated health checking, Consul can route traffic away from unhealthy nodes, allowing systems and services to gracefully recover. In addition, Consul’s service discovery layer works with Terraform. Consul-Terraform-Sync (CTS) automates updates to network infrastructure based on dynamic changes to each service. For example, as services scale up or down, CTS can trigger Terraform to update firewalls or load balancers to reflect the latest changes. Also, since each datacenter runs independently, supporting multiple datacenters is no different than supporting a single datacenter.
```

> Consul is not a replacement for other configuration management tools. These tools are still critical for setting up applications, including Consul. Static provisioning is best managed by existing tools while dynamic state and service discovery is better managed by Consul. 

Reviewer 2: ```suggestion
Consul is not a replacement for other configuration management tools. These tools are still critical for setting up applications, including Consul. Static provisioning is best managed by existing tools, while Consul enables you to leverage dynamic configuration and service discovery. 
```
